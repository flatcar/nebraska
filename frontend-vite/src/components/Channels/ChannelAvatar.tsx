import { Theme } from '@mui/material';
import Avatar from '@mui/material/Avatar';
import makeStyles from '@mui/styles/makeStyles';

export interface ChannelAvatarProps {
  backgroundColor?: string;
  color?: string;
  size?: string | number;
  children?: React.ReactNode;
}

const useStyles = makeStyles((theme: Theme) => ({
  colorAvatar: (props: ChannelAvatarProps) => ({
    color: 'rgb(15 15 15)',
    backgroundColor: props.backgroundColor || props.color || theme.palette.secondary.main,
    width: props.size,
    height: props.size,
    display: 'inline-flex',
  }),
}));

export default function ChannelAvatar(props: ChannelAvatarProps) {
  const classes = useStyles(props);

  return <Avatar className={classes.colorAvatar}>{props.children || ' '}</Avatar>;
}
