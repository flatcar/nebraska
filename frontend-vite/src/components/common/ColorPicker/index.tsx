export { default } from './ColorPicker';
export * from './ColorPicker';
