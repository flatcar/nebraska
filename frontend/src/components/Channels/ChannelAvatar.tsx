import Avatar from '@mui/material/Avatar';

export interface ChannelAvatarProps {
  backgroundColor?: string;
  color?: string;
  size?: string | number;
  children?: React.ReactNode;
}

export default function ChannelAvatar({
  backgroundColor,
  color,
  size,
  children,
}: ChannelAvatarProps) {
  return (
    <Avatar
      sx={{
        color: 'rgb(15 15 15)',
        display: 'inline-flex',
        backgroundColor: theme => backgroundColor ?? color ?? theme.palette.secondary.main,
        width: size,
        height: size,
      }}
    >
      {children || ' '}
    </Avatar>
  );
}
