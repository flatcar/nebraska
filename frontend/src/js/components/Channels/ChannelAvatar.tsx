import { Theme } from '@material-ui/core';
import Avatar from '@material-ui/core/Avatar';
import makeStyles from '@material-ui/styles/makeStyles';
import React from 'react';

interface ChannelAvatarProps {
  backgroundColor?: string;
  color?: string;
  size?: number;
  children?: React.ReactNode;
}

const useStyles = makeStyles( (theme: Theme) => ({
  colorAvatar: (props: ChannelAvatarProps) => ({
    color: '#fff',
    backgroundColor: props.backgroundColor || props.color ||
    theme.palette.secondary.main,
    width: props.size,
    height: props.size,
    display: 'inline-flex'
  }),
}));

export default function ChannelAvatar(props: ChannelAvatarProps) {
  const classes = useStyles(props);

  return (
    <Avatar className={classes.colorAvatar}>{props.children || '' }</Avatar>
  );
}
