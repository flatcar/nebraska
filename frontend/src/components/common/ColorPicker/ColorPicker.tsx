import IconButton from '@mui/material/IconButton';
import Popover from '@mui/material/Popover';
import { makeStyles } from '@mui/styles';
import React from 'react';
import { TwitterPicker } from 'react-color';

import ChannelAvatar from '../../Channels/ChannelAvatar';

// @todo: This needs to become a FormControl so we can display it in a similar
// style as the other form controls.

const useStyles = makeStyles({
  iconButton: {
    padding: '0',
  },
});

export interface ColorPickerProps {
  /** Text color, applied to the children and used in the popover. */
  color: string;
  /** When a color is chosen. */
  onColorPicked: (color: { hex: string }) => void;
  /** Should the color picker be displayed initially? */
  initialOpen?: boolean;
  children: React.ReactElement<any>;
}

export default function ColorPicker(props: ColorPickerProps) {
  const classes = useStyles();
  const [channelColor, setChannelColor] = React.useState(props.color);
  const [displayColorPicker, setDisplayColorPicker] = React.useState(props.initialOpen);
  const [anchorEl, setAnchorEl] = React.useState(null);
  const { onColorPicked } = props;

  function handleColorChange(color: { hex: string }) {
    setChannelColor(color.hex);
    onColorPicked(color);
  }

  function handleColorButtonClick(event: any) {
    setAnchorEl(event.currentTarget);
    setDisplayColorPicker(true);
  }

  function handleClose() {
    setAnchorEl(null);
    setDisplayColorPicker(false);
  }

  return (
    <div>
      <IconButton
        className={classes.iconButton}
        onClick={handleColorButtonClick}
        data-testid="icon-button"
        size="large"
      >
        {props.children ? (
          <ChannelAvatar color={channelColor}>{props.children}</ChannelAvatar>
        ) : (
          <ChannelAvatar color={channelColor} />
        )}
      </IconButton>
      {displayColorPicker && anchorEl && (
        <Popover
          data-testid="popover"
          open={displayColorPicker}
          anchorEl={anchorEl}
          onClose={handleClose}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'center',
          }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'center',
          }}
        >
          <TwitterPicker
            color={channelColor}
            onChangeComplete={handleColorChange}
            triangle="hide"
          />
        </Popover>
      )}
    </div>
  );
}
