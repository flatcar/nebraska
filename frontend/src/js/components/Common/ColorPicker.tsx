import IconButton from '@material-ui/core/IconButton';
import Popover from '@material-ui/core/Popover';
import { makeStyles } from '@material-ui/styles';
import React from 'react';
import { TwitterPicker } from 'react-color';
import ChannelAvatar from '../Channels/ChannelAvatar';

// @todo: This needs to become a FormControl so we can display it in a similar
// style as the other form controls.

const useStyles = makeStyles({
  iconButton: {
    padding: '0',
  },
});

interface ColorPickerButtonProps {
  color: string;
  onColorPicked: (color: {hex: string}) => void;
  componentColorProp: string;
  children: React.ReactElement<any>;
}

export function ColorPickerButton(props: ColorPickerButtonProps) {
  const classes = useStyles();
  const [channelColor, setChannelColor] = React.useState(props.color);
  const [displayColorPicker, setDisplayColorPicker] = React.useState(false);
  const [anchorEl, setAnchorEl] = React.useState(null);
  const {onColorPicked, componentColorProp} = props;

  const componentProps: {[key: string]: string} = {};
  componentProps[componentColorProp] = channelColor;

  function handleColorChange(color: {hex: string}) {
    setChannelColor(color.hex);
    onColorPicked(color);
  }

  function handleColorButtonClick(event: any) {
    setAnchorEl(event.currentTarget);
    setDisplayColorPicker(true);
  }

  function handleClose() {
    setAnchorEl(null);
    setDisplayColorPicker(false);
  }

  return (
    <div>
      <IconButton className={classes.iconButton}
        onClick={handleColorButtonClick}
        data-testid="icon-button"
      >
        {props.children ?
          React.cloneElement(props.children as React.ReactElement<any>, componentProps)
          :
          <ChannelAvatar color={channelColor} />
        }
      </IconButton>
      {displayColorPicker &&
      <Popover
        data-testid="popover"
        open={displayColorPicker}
        anchorEl={anchorEl}
        onClose={handleClose}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
      >
        <TwitterPicker color={channelColor} onChangeComplete={handleColorChange} triangle="hide"/>
      </Popover>
      }
    </div>
  );
}
