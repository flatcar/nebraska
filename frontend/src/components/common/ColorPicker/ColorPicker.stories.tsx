import { Meta, StoryObj } from '@storybook/react';

import ChannelAvatar from '../../Channels/ChannelAvatar';
import ColorPicker from './ColorPicker';

const meta: Meta<typeof ColorPicker> = {
  title: 'ColorPickerButton',
  component: ColorPicker,
  argTypes: {
    onColorPicked: { action: 'onColorPicked' },
  },
};
export default meta;

type Story = StoryObj<typeof ColorPicker>;

export const Closed: Story = {
  args: {
    color: '#EB144C',
    children: <ChannelAvatar>Beta</ChannelAvatar>,
    initialOpen: false,
  },
};

export const Open: Story = {
  args: {
    color: '#EB144C',
    children: <ChannelAvatar>Beta</ChannelAvatar>,
    initialOpen: true,
  },
};
