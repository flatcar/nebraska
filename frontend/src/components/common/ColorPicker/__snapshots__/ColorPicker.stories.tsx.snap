// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPickerButton Closed smoke-test 1`] = `
<div class="css-1t32mh7">
  <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge ColorPicker-iconButton css-154kngp-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
          data-testid="icon-button"
  >
    <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1mo2r65-MuiAvatar-root">
      <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1yqfavt-MuiAvatar-root">
        Beta
      </div>
    </div>
  </button>
</div>
`;

exports[`ColorPickerButton Open smoke-test 1`] = `
<div class="css-1t32mh7">
  <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge ColorPicker-iconButton css-154kngp-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
          data-testid="icon-button"
  >
    <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1mo2r65-MuiAvatar-root">
      <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1yqfavt-MuiAvatar-root">
        Beta
      </div>
    </div>
  </button>
</div>
`;
